### PR TITLE
network: Allow to configure VLan egress qos maps

### PR DESCRIPTION
**Pull request by ssahani**
Closes https://github.com/systemd/systemd/issues/17989

### CODE REVIEW

File: man/systemd.netdev.xml [hunk 1]
```diff
@@ -432,6 +432,15 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>EgressQOSMaps=</varname></term>
+        <listitem>
+          <para>The EgressQOSMaps defines a mapping of Linux internal packet priority (SO_PRORITY) to VLAN header
+          PCP field for outgoing frames. A whitespace-separated list unsigned integers in the format from-to.
```
Comment (bluca) [inline man/systemd.netdev.xml]: "list of"

Would be good to have an example, given it's not a straightforward single-value directive.

Comment (ssahani) [reply man/systemd.netdev.xml]: Yes right will update thanks

File: src/network/netdev/vlan.c [hunk 2]
```diff
@@ -54,6 +55,96 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+
+                r = parse_range(w, &m->from, &m->to);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
+                                   "Failed to parse %s, ignoring: %s", lvalue, w);
+                        return r;
+                }
+
+                if (m->to > m->from) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
```
Comment (bluca) [inline src/network/netdev/vlan.c]: if the range can't be parsed the error is returned, but if the value here is wrong 0 is returned - shouldn't it be consistent for parsing errors?

Comment (ssahani) [reply src/network/netdev/vlan.c]: should we add ``` m->to == 0 && m->from ==0``` ?

File: src/network/netdev/vlan.c [hunk 3]
```diff
@@ -54,6 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
 
+        if (!set_isempty(v->egress_qos_maps)) {
+                struct ifla_vlan_qos_mapping *m;
+
+                r = sd_netlink_message_open_container(req, IFLA_VLAN_EGRESS_QOS);
+                if (r < 0)
+                        return log_netdev_error_errno(netdev, r, "Could not open container IFLA_VLAN_EGRESS_QOS: %m");
+
+                SET_FOREACH(m, v->egress_qos_maps) {
+                        r = sd_netlink_message_append_data(req, IFLA_VLAN_QOS_MAPPING, m, sizeof(struct ifla_vlan_qos_mapping));
```
Comment (yuwata) [inline src/network/netdev/vlan.c:66]: ~~If I remember correctly, `sd_netlink_message_append_data()` overwrites the previous assignment. So, if the set contains multiple elements, this does not work correctly...~~

File: src/network/netdev/vlan.c [hunk 4]
```diff
@@ -54,6 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                        if (r < 0)
+                                return log_netdev_error_errno(netdev, r, "Could not append IFLA_VLAN_QOS_MAPPING attribute: %m");
+
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: Please drop this unnecessary blank line.

File: src/network/netdev/vlan.c [hunk 5]
```diff
@@ -54,6 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                r = sd_netlink_message_open_container(req, IFLA_VLAN_INGRESS_QOS);
+                if (r < 0)
+                        return log_netdev_error_errno(netdev, r, "Could not open container IFLA_VLAN_INGRESS_QOS: %m");
+
+                SET_FOREACH(m, v->ingress_qos_maps) {
+                        r = sd_netlink_message_append_data(req, IFLA_VLAN_QOS_MAPPING, m, sizeof(struct ifla_vlan_qos_mapping));
+                        if (r < 0)
+                                return log_netdev_error_errno(netdev, r, "Could not append IFLA_VLAN_QOS_MAPPING attribute: %m");
+
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: Here too.

File: src/network/netdev/vlan.c [hunk 6]
```diff
@@ -54,6 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                        return 0;
+
+                m = new0(struct ifla_vlan_qos_mapping, 1);
+                if (!m)
+                        return log_oom();
+
+                r = parse_range(w, &m->from, &m->to);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse %s, ignoring: %s", lvalue, w);
+                        return 0;
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: I prefer not to drop remaining elements, that is, `return 0;` -> `continue;`.

File: src/network/netdev/vlan.c [hunk 7]
```diff
@@ -54,6 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                }
+
+                if ((m->to > m->from) || (m->to == 0 && m->from == 0)) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Invalid %s, ignoring: %s", lvalue, w);
+                        return 0;
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: Same here. `return 0;` -> `continue;`

File: src/network/netdev/vlan.c [hunk 8]
```diff
@@ -54,6 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                }
+
+                r = set_ensure_put(s, NULL, m);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to store %s, ignoring: %s", lvalue, w);
+                        return 0;
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: Also here. `return 0;` -> `continue;`

File: src/network/netdev/vlan.c [hunk 9]
```diff
@@ -54,6 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse %s, ignoring: %s", lvalue, w);
+                        return 0;
+                }
+
+                if ((m->to > m->from) || (m->to == 0 && m->from == 0)) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Invalid %s, ignoring: %s", lvalue, w);
+                        return 0;
+                }
+
+                r = set_ensure_put(s, NULL, m);
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: I prefer to dedup elements here. So, please introduce an specific `hash_ops` for this.

Also, please use `set_ensure_consume(s, &some_specific_hash_ops, TAKE_PTR(m));`

**Review by yuwata — CHANGES_REQUESTED** [review]

File: man/systemd.netdev.xml [hunk 10]
```diff
@@ -432,6 +432,24 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>EgressQOSMaps=</varname></term>
+        <listitem>
+          <para>The EgressQOSMaps defines a mapping of Linux internal packet priority (SO_PRORITY) to VLAN header
+          PCP field for outgoing frames. A whitespace-separated list of unsigned integers in the format from-to i.e.
+          21:15 45:5. When unset, the kernel's default will be used.
```
Comment (yuwata) [inline man/systemd.netdev.xml]: Please use `<literal>` tag for the example. Also mention that `from` must be larger than or equal to `to`.

**Comment by ssahani** [comment]
Updated thanks for the review. ptal

File: man/systemd.netdev.xml [hunk 11]
```diff
@@ -432,6 +432,27 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>EgressQOSMaps=</varname></term>
+        <listitem>
+          <para>The EgressQOSMaps defines a mapping of Linux internal packet priority (SO_PRORITY) to VLAN header
```
Comment (yuwata) [inline man/systemd.netdev.xml]: `The EgressQOSMaps defines` -> `This defines` or `Defines`.

`SO_PRORITY` -> `SO_PRIORITY`.

File: man/systemd.netdev.xml [hunk 12]
```diff
@@ -432,6 +432,27 @@
...
+          PCP field for outgoing frames. A whitespace-separated list of unsigned integers in the format
+          <literal>from</literal>-<literal>to</literal> i.e. 21:15 45:5. Note that <literal>from</literal> must be greater
```
Comment (yuwata) [inline man/systemd.netdev.xml]: `Takes a whitespace-separated list of unsigned integer pairs in the format <literal>from</literal>:<literal>to</literal>, e.g., <literal>21:15 45:5</literal>`.

File: man/systemd.netdev.xml [hunk 13]
```diff
@@ -432,6 +432,27 @@
...
+          than or equal to <literal>to</literal>. When unset, the kernel's default will be used.
+          </para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>IngressQOSMaps=</varname></term>
+        <listitem>
+          <para>The IngressQOSMaps defines a mapping of Linux internal packet priority (SO_PRORITY) to VLAN header
```
Comment (yuwata) [inline man/systemd.netdev.xml]: Same here.

- `This defines ...` or `Defines ...`.
- SO_PRORITY -> SO_PRIORITY

File: src/network/netdev/vlan.c [hunk 14]
```diff
@@ -54,6 +55,129 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+        Set **s = data;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (isempty(rvalue)) {
+                *s = set_free_free(*s);
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: Please use `set_free()`, as the hash_ops already specified `free`.

File: src/network/netdev/vlan.c [hunk 15]
```diff
@@ -54,6 +55,129 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                        return log_oom();
+
+                r = parse_range(w, &m->from, &m->to);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse %s, ignoring: %s", lvalue, w);
+                        continue;
+                }
+
+                if ((m->to > m->from) || (m->to == 0 && m->from == 0)) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Invalid %s, ignoring: %s", lvalue, w);
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: `r` -> `0`.

File: src/network/netdev/vlan.c [hunk 16]
```diff
@@ -54,6 +55,129 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                        continue;
+                }
+
+                r = set_ensure_put(s, &vlan_qos_maps_hash_ops, TAKE_PTR(m));
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: `set_ensure_put()` -> `set_ensure_consume()`.

File: src/network/netdev/vlan.c [hunk 17]
```diff
@@ -54,6 +55,129 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to store %s, ignoring: %s", lvalue, w);
+                        continue;
+
+                }
+        }
+
         return 0;
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: Please drop unnecessary `return 0;`, as the above `for` loop does not break.

File: src/network/netdev/vlan.c [hunk 18]
```diff
@@ -75,6 +199,17 @@ static int netdev_vlan_verify(NetDev *netdev, const char *filename) {
...
 
+static void vlan_done(NetDev *n) {
+        VLan *v;
+
+        v = VLAN(n);
+
+        assert(v);
+
+        set_free_free(v->egress_qos_maps);
+        set_free_free(v->ingress_qos_maps);
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: Please use `set_free()`.

**Review by yuwata — CHANGES_REQUESTED** [review]
Thank you. Looks mostly OK. One more round please.

File: man/systemd.netdev.xml (hunk 11 above)
Comment (ssahani) [reply man/systemd.netdev.xml]: Specifies also nice.

File: man/systemd.netdev.xml (hunk 12 above)
Comment (ssahani) [reply man/systemd.netdev.xml]: `parse_range` takes in the format `a-b`

File: test/fuzz/fuzz-netdev-parser/directives.netdev [hunk 19]
```diff
@@ -5,6 +5,7 @@ ReorderHeader=
 Id=
 GVRP=
 Protocol=
+EgressQOSMaps=
```
Comment (yuwata) [inline test/fuzz/fuzz-netdev-parser/directives.netdev:8]: `IngressQOSMaps=` is missing.

File: man/systemd.netdev.xml [hunk 20]
```diff
@@ -443,6 +443,18 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
```
Comment (yuwata) [inline man/systemd.netdev.xml]: Also, could you fix the commit message of the third commit? `egress` -> `ingress`.

File: test/fuzz/fuzz-netdev-parser/directives.netdev (hunk 19 above)
Comment (ssahani) [reply test/fuzz/fuzz-netdev-parser/directives.netdev:8]: ouch. thanks updated .

**Comment by yuwata** [comment]
Looks perfect! Thank you. Let's wait for the CIs.

**Review by yuwata — APPROVED** [review]

File: man/systemd.netdev.xml [hunk 21]
```diff
@@ -432,6 +432,17 @@
...
+        <term><varname>EgressQOSMaps=</varname></term>
+        <listitem>
+          <para>Defines a mapping of Linux internal packet priority (SO_PRIORITY) to VLAN header
```
Comment (poettering) [inline man/systemd.netdev.xml]: doesn't matter, but maybe enclose `SO_PRIORITY` in `<constant></constant>` or so?

File: man/systemd.netdev.xml [hunk 22]
```diff
@@ -432,6 +432,17 @@
...
+          PCP field for outgoing frames. Takes a whitespace-separated list of unsigned integer pairs
+          in the format <literal>from</literal>-<literal>to</literal>, e.g., <literal>21-15 45-5</literal>.
+          Note that <literal>from</literal> must be greater than or equal to <literal>to</literal>. When unset,
+          the kernel's default will be used.
```
Comment (poettering) [inline man/systemd.netdev.xml]: please specify the valid range of the integer pairs.

File: src/network/netdev/vlan.c [hunk 23]
```diff
@@ -54,9 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                        return 0;
+                }
+                if (r == 0)
+                        return 0;
+
+                m = new0(struct ifla_vlan_qos_mapping, 1);
+                if (!m)
+                        return log_oom();
+
+                r = parse_range(w, &m->from, &m->to);
```
Comment (poettering) [inline src/network/netdev/vlan.c:162]: is really the full range of `unsigned` OK here?

File: src/network/netdev/vlan.c [hunk 24]
```diff
@@ -54,9 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse %s, ignoring: %s", lvalue, w);
+                        continue;
+                }
+
+                if ((m->to > m->from) || (m->to == 0 && m->from == 0)) {
```
Comment (poettering) [inline src/network/netdev/vlan.c]: this doesn't look right... you refuse things if to is larger than from? you probably mean the reverse...

also, is 0-0 forbidden, but 0-5 shall be allowed? really?

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/network/netdev/vlan.c (hunk 23 above)
Comment (ssahani) [reply src/network/netdev/vlan.c:162]: yes ```11: vlan1@dummy98: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 0a:f9:c4:b8:5d:54 brd ff:ff:ff:ff:ff:ff promiscuity 0 minmtu 0 maxmtu 65535 
    vlan protocol 802.1Q id 23 <REORDER_HDR> 
      egress-qos-map { 4294967294:2 } addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535  ```

File: src/network/netdev/vlan.c (hunk 24 above)
Comment (ssahani) [reply src/network/netdev/vlan.c]: No it's in the format from-to, so 0-5 is not allowed . ```from >= to``` also it supports full range .  ```0``` either from or to is not supported.

File: man/systemd.netdev.xml [hunk 25]
```diff
@@ -432,7 +432,7 @@
           </para>
         </listitem>
       </varlistentry>
-    </variablelist>
+   </variablelist>
```
Comment (poettering) [inline man/systemd.netdev.xml]: hmm, changing from 2 space indentation to 1 space?

File: src/network/netdev/vlan.c (hunk 24 above)
Comment (poettering) [reply src/network/netdev/vlan.c]: hm, but that's not what this checks...

do you mean:
```c
if (m->to > m->from || m->to == 0 || m->from)
…
```
???

all examples for vlan egress qos maps have a lower value first, higher value second, hence the check you have here seems reversed... i don#t get this

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/network/netdev/vlan.c (hunk 24 above)
Comment (ssahani) [reply src/network/netdev/vlan.c]: ```<literal>21-15 45-5</literal>``` . Multiple values is allowed  `21-15` is one set and `45-5` is another. 


```
[VLAN]
Id=23
EgressQOSMaps=21-7 45-5

```
results in 
```

 egress-qos-map { 21:7 45:5 } addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 
                                                                                                                       
```

I just revered the check which is same.

let me add 
```
/usr/lib/systemd/network/vlan.netdev:7: Invalid EgressQOSMaps, ignoring: 21-0
/usr/lib/systemd/network/vlan.netdev:7: Invalid EgressQOSMaps, ignoring: 5-7
```

File: src/network/netdev/vlan.c [hunk 26]
```diff
@@ -54,9 +55,111 @@ static int netdev_vlan_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+                if (!m)
+                        return log_oom();
+
+                r = parse_range(w, &m->from, &m->to);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse %s, ignoring: %s", lvalue, w);
+                        continue;
+                }
+
+                if ((m->to > m->from) || (m->to == 0 || m->from == 0)) {
```
Comment (yuwata) [inline src/network/netdev/vlan.c]: Then, all `()` are redundant:
`if (m->to > m->from || m->to == 0 || m->from == 0) {`

**Pull request merged**